### PR TITLE
feat: make full precision decentralized op stateless

### PR DESCRIPTION
**Pull request by wangraying**
BREAKING CHANGE:

`BaguaBucket.append_decentralized_synchronous_op` now only supports full precision decentralized communication.

### CODE REVIEW

File: bagua/torch_api/algorithms/decentralized.py [hunk 1]
```diff
@@ -54,22 +56,47 @@ def hook(parameter_name, parameter):
...
         bucket: BaguaBucket,
     ):
+        self._init_states(bucket)
+        torch.cuda.synchronize()
         bucket.clear_ops()
         bucket.append_decentralized_synchronous_op(
             hierarchical=self.hierarchical,
             peer_selection_mode=self.peer_selection_mode,
             communication_interval=self.communication_interval,
+            left_peer_weight=bucket._peer_weight,
```
Comment (NOBLES5E) [inline bagua/torch_api/algorithms/decentralized.py:99]: weird to call it `left_peer_weight`

File: bagua/torch_api/algorithms/decentralized.py [hunk 2]
```diff
@@ -54,22 +56,47 @@ def hook(parameter_name, parameter):
...
+                        broadcast(bucket.backend_tensor, 0, intra_comm)
+
+            def copyback_worker_fn(*unused):
+                for bucket in bagua_module.bagua_buckets:
+                    if self.hierarchical:
+                        broadcast(bucket.backend_tensor, 0, intra_comm)
+                    else:
+                        bucket.backend_tensor.copy_(bucket._peer_weight)
+
+            bagua_module._bagua_backend.schedule_python_op(
```
Comment (NOBLES5E) [inline bagua/torch_api/algorithms/decentralized.py:76]: just run `copyback_xxx_fn`?

File: bagua/torch_api/algorithms/decentralized.py (hunk 1 above)
Comment (NOBLES5E) [reply bagua/torch_api/algorithms/decentralized.py:99]: we can create two separate ops for low precision decentralized and decentralized

File: bagua/torch_api/bucket.py [hunk 3]
```diff
@@ -260,14 +244,63 @@ def append_decentralized_synchronous_op(
...
+        Returns:
+            The bucket itself.
+        """
+
+        if hierarchical:
+            self.backend_bucket.append_low_precision_decentralized_synchronous_op(
+                self._bagua_backend.internode_communicator,
+                self._bagua_backend.intranode_communicator,
+                hierarchical=hierarchical,
+                peer_selection_mode="ring",
```
Comment (NOBLES5E) [inline bagua/torch_api/bucket.py:286]: we don't need to specify ring anymore?

File: bagua/torch_api/algorithms/decentralized.py [hunk 4]
```diff
@@ -29,6 +30,7 @@ def __init__(
         self.hierarchical = hierarchical
         self.peer_selection_mode = peer_selection_mode
         self.communication_interval = communication_interval
+        self.step = 0
```
Comment (NOBLES5E) [inline bagua/torch_api/algorithms/decentralized.py:33]: where do we need this step?

File: bagua/torch_api/bucket.py [hunk 5]
```diff
@@ -218,40 +215,27 @@ def append_decentralized_synchronous_op(
         when all the tensors within the bucket are marked ready.
 
         Args:
+            peer_weight (BaguaTensor): A tensor used for receiving and averaging model with peers, should be the same size
```
Comment (NOBLES5E) [inline bagua/torch_api/bucket.py:218]: we need better explanation on how this operator works, since it is not an inplace operation. The result will be in `peer_weight`. Users (who develop new algorithms) need to understand how this algorithm actually works and what they need to do after using this op

File: bagua/torch_api/bucket.py [hunk 6]
```diff
@@ -218,40 +215,27 @@ def append_decentralized_synchronous_op(
...
-            left_peer_weight (BaguaTensor): Model replica of current worker's connected left peer, a flattened tensor containing
-                the same data as model weights of left peer, required for low precision decentralized operation.
-            right_peer_weight (BaguaTensor): Model replica of current worker's connected right peer, similarly as `left_peer_weight`,
-                required for low precision decentralized operation.
         Returns:
             The bucket itself.
         """
+
         if hierarchical:
             self.backend_bucket.append_decentralized_synchronous_op(
```
Comment (NOBLES5E) [inline bagua/torch_api/bucket.py:232]: decentralize_op = ...
decentralize_op.copy_back_peer_weight()

File: bagua/torch_api/algorithms/decentralized.py [hunk 7]
```diff
@@ -53,23 +55,40 @@ def hook(parameter_name, parameter):
...
+
     def init_operations(
         self,
         bagua_module: BaguaModule,
         bucket: BaguaBucket,
     ):
+        self._init_states(bucket)
+        torch.cuda.synchronize()
         bucket.clear_ops()
         bucket.append_decentralized_synchronous_op(
```
Comment (NOBLES5E) [inline bagua/torch_api/algorithms/decentralized.py:87]: bucket.decentralize_op = bucket.append_decentralized_synchronous_op(...)
bucket.decentralize_op.copy_back_peer_weight()

File: bagua/torch_api/algorithms/decentralized.py [hunk 8]
```diff
@@ -53,23 +55,40 @@ def hook(parameter_name, parameter):
...
             hierarchical=self.hierarchical,
             peer_selection_mode=self.peer_selection_mode,
             communication_interval=self.communication_interval,
```
Comment (NOBLES5E) [inline bagua/torch_api/algorithms/decentralized.py:90]: remove communication op, only mark tensor ready when `bagua_train_step_counter` % comm_interval == 0

Comment (wangraying) [reply bagua/torch_api/algorithms/decentralized.py]: > remove communication interval, only mark tensor ready when `bagua_train_step_counter` % comm_interval == 0

fixed, add a new method `_bagua_not_skip_this_step` in `BaguaModule`

File: bagua/torch_api/algorithms/decentralized.py [hunk 9]
```diff
@@ -54,38 +51,43 @@ def hook(parameter_name, parameter):
...
         bucket: BaguaBucket,
     ):
+        self._init_states(bucket)
+        torch.cuda.synchronize()
         bucket.clear_ops()
         bucket.append_decentralized_synchronous_op(
+            peer_weight=bucket._peer_weight,
             hierarchical=self.hierarchical,
             peer_selection_mode=self.peer_selection_mode,
-            communication_interval=self.communication_interval,
```
Comment (NOBLES5E) [inline bagua/torch_api/algorithms/decentralized.py:72]: we removed communication interval support?

File: bagua/torch_api/distributed.py [hunk 10]
```diff
@@ -147,6 +147,7 @@ def with_bagua(  # pytype: disable=module-attr
         self,
         optimizers: List[torch.optim.Optimizer],
         algorithm: "bagua.torch_api.algorithms.Algorithm",
+        communication_interval: int = 1,
```
Comment (NOBLES5E) [inline bagua/torch_api/distributed.py:150]: this should be in the algorithm, since it does not make sense for some algorithms like gradient allreduce but tends to get misused.

File: bagua/torch_api/distributed.py [hunk 11]
```diff
@@ -199,7 +201,9 @@ def with_bagua(  # pytype: disable=module-attr
             self, "_ddp_params_and_buffers_to_ignore"
         ):  # for compatibility with PyTorch DDP
             self.parameters_to_ignore.extend(self._ddp_params_and_buffers_to_ignore)
-        self.bagua_train_step_counter = 0
+
+        self.communication_interval = communication_interval
+        self.bagua_train_step_counter = -1
```
Comment (NOBLES5E) [inline bagua/torch_api/distributed.py:206]: why start with -1

File: bagua/torch_api/distributed.py [hunk 12]
```diff
@@ -386,10 +392,15 @@ def real_post_backward_hook(*unused):
             if not hasattr(optimizer, "_bagua_original_step"):
                 optimizer._bagua_original_step = optimizer.step
 
+            if not hasattr(optimizer, "_bagua_module"):
+                optimizer._bagua_module = self
```
Comment (NOBLES5E) [inline bagua/torch_api/distributed.py:396]: why do we need optimizer having the module attribute

**Review by NOBLES5E — CHANGES_REQUESTED** [review]
see comments.

avoid changes to `bagua/torch_api/distributed.py` unless necessary

File: bagua/torch_api/bucket.py [hunk 13]
```diff
@@ -203,71 +204,129 @@ def append_centralized_synchronous_op(
...
+        In this operation, local model weights are copied into `peer_weight`, and each worker sends `peer_weight` to its
+        peer and receives its peer's counterpart, then average them. Since the original local model weights remains
+        unchanged during this process, it is possible to overlap communication with the computation of `forward` and
+        `backward` process.
+
+        This function initiates the above mentioned process, which will be executed by the Bagua backend in
+        the order they are appended when all the tensors within the bucket are marked ready.
+
+        To complete this operation, it is needed to copy `peer_weight` back to local model weights once `backward` finishes.
+        This is done by another function, see :func:`decentralized_synchronous_op_copy_back_peer_weight`.
```
Comment (NOBLES5E) [inline bagua/torch_api/bucket.py:223]: In this operation, local model weights are copied into `self.peer_weight`, and the result of decentralized averaging will be in `self.peer_weight`.
        To copy `self.peer_weight` back to self, call `self.decentralized_synchronous_op_copy_back_peer_weight()`.

File: bagua/torch_api/distributed.py (hunk 11 above)
Comment (wangraying) [reply bagua/torch_api/distributed.py:206]: > why start with -1

this `bagua_train_step_counter` is increased in pre-forward hook, initialize it to be -1, to make sure the first step will not be skipped.

File: bagua/torch_api/distributed.py (hunk 10 above)
Comment (wangraying) [reply bagua/torch_api/distributed.py:150]: oh. it sounds better.

File: bagua/torch_api/distributed.py (hunk 12 above)
Comment (wangraying) [reply bagua/torch_api/distributed.py:396]: to skip optimizer step hook.

File: bagua/torch_api/bucket.py (hunk 13 above)
Comment (wangraying) [reply bagua/torch_api/bucket.py:223]: > ```
>     This operation is not inplace, which means the bucket content is first copied to `self.peer_weight`, and the result of decentralized averaging will be in `self.peer_weight`.
>     To copy `self.peer_weight` back to self, call `self.decentralized_synchronous_op_copy_back_peer_weight()`.
> ```
It we implement it as described above, `peer_weight` in the argument list can be removed

Comment (NOBLES5E) [reply bagua/torch_api/bucket.py:223]: Oh we don't want to do hardcode self.peer_weight

```
This operation is not inplace, which means the bucket content is first copied to `peer_weight`, and the result of decentralized averaging will be in `peer_weight`. To copy `peer_weight` back to self, call `bucket.decentralized_synchronous_op_copy_back_peer_weight(...)`.
```

File: bagua/torch_api/distributed.py (hunk 11 above)
Comment (NOBLES5E) [reply bagua/torch_api/distributed.py:206]: that sound hacky.

Let's just remove changes to this file and move all communication interval logic to the algorithm itself

File: bagua/torch_api/bucket.py [hunk 14]
```diff
@@ -203,71 +204,125 @@ def append_centralized_synchronous_op(
...
+        """
+        Copy `peer_weight` back to bucket weights to end a decentralized synchronous operation.
+        See :func:`append_decentralized_synchronous_op` for more information.
+
+        Args:
+            peer_weight (BaguaTensor):  A tensor used for averaging model with peers, should be the same size
+                with bucket weights, i.e. `self` tensors and padding tensor (if exists).
+            hierarchical (bool): Enable hierarchical communication. Which means the GPUs on the same machine
+                will communicate will each other first. After that, machines do inter-node communication. This can
+                boost performance when the inter-node communication cost is high.
```
Comment (NOBLES5E) [inline bagua/torch_api/bucket.py:264]: this must be the same with append_decentralized_synchronous_op's hierarchical argument?

File: bagua/torch_api/bucket.py [hunk 15]
```diff
@@ -203,71 +204,127 @@ def append_centralized_synchronous_op(
...
+
+        The operations will be executed by the Bagua backend in the order they are appended
+        when all the tensors within the bucket are marked ready.
+
+        Args:
+            weight (BaguaTensor): Model replica of current worker's local model. It should be of the same size
+                with the bucket tensors total size. Use ``self.flattened_tensor`` to create such a tensor.
+            left_peer_weight (BaguaTensor): Model replica of current worker's left peer. It should be of the same size
+                with the bucket tensors total size. Use ``self.flattened_tensor`` to create such a tensor.
+                should contain the same data as initializing weights of current worker's left peer.
```
Comment (NOBLES5E) [inline bagua/torch_api/bucket.py:295]: is this automatic or the user need to first get peer's weight?

File: bagua/torch_api/bucket.py [hunk 16]
```diff
@@ -203,71 +204,127 @@ def append_centralized_synchronous_op(
...
+            right_peer_weight (BaguaTensor): Model replica of current worker's right peer, It should be of the same size
+                with the bucket tensors total size. Use ``self.flattened_tensor`` to create such a tensor. 
+                should contain the same data as initializing weights of current worker's right peer.
```
Comment (NOBLES5E) [inline bagua/torch_api/bucket.py:298]: is this automatic or the user need to first get peer's weight?

**Review by NOBLES5E — CHANGES_REQUESTED** [review]
see comments

File: bagua/torch_api/algorithms/decentralized.py [hunk 17]
```diff
@@ -40,8 +41,9 @@ def init_tensors(self, bagua_module: BaguaModule) -> List[BaguaTensor]:
 
     def init_forward_pre_hook(self, bagua_module: BaguaModule):
         def hook(input):
-            for tensor in self.tensors:
-                tensor.bagua_mark_communication_ready()
+            if bagua_module.bagua_train_step_counter % self.communication_interval == 0:
```
Comment (NOBLES5E) [inline bagua/torch_api/algorithms/decentralized.py:44]: just change this to if _should_communicate(...)
